src/java/picard/sam/markduplicates/MarkDuplicates.java

```diff
@@ -443,18 +506,27 @@ private void generateDuplicateIndexes() {
         this.duplicateIndexes.doneAddingStartIteration();
     }
 
-    private boolean areComparableForDuplicates(final ReadEndsForMarkDuplicates lhs, final ReadEndsForMarkDuplicates rhs, final boolean compareRead2) {
-        boolean retval = lhs.libraryId == rhs.libraryId &&
-                lhs.read1ReferenceIndex == rhs.read1ReferenceIndex &&
-                lhs.read1Coordinate == rhs.read1Coordinate &&
-                lhs.orientation == rhs.orientation;
+    private boolean areComparableForDuplicates(final ReadEndsForMarkDuplicates lhs, final ReadEndsForMarkDuplicates rhs, final boolean compareRead2, final boolean useBarcodes) {
+        boolean areComparable = lhs.libraryId == rhs.libraryId;
+
+        if (areComparable && useBarcodes) {
+            if (areComparable) areComparable = 0 == compareInteger(((ReadEndsForMarkDuplicatesWithBarcodes) lhs).barcode, ((ReadEndsForMarkDuplicatesWithBarcodes) rhs).barcode);
```

Comment:
given that you only want ==0 i.e are equal here, why not use equality of the barcodes?